README
rule30
======

Version 0.6.1 (October 9, 2016)

rule30 is a collection of functions that implement Wolfram's 
[Rule 30](http://en.wikipedia.org/wiki/Rule_30) cellular automaton and related 
transformations.

The Algorithm
-------------

Rule 30 is a one-dimensional two-state cellular automaton rule. The automaton 
consists of a sequence of symbols (sometimes thought of as a row of cells). Two 
states mean that the symbol can take one of two values (0 or 1), or, in other 
words, the corrssponding cell can be either "dead" or "alive". The state of the 
sequence is computed from the previous state. The state of a cell depends on 
the previous state of the cell itself, as well as the two neighbouring cells.

The rule is:

s'(i) = s(i-1) XOR [ s(i) OR s(i+1) ]

All possible inputs and outcomes are summarized below:

| current state | 000 | 001 | 010 | 011 | 100 | 101 | 110 | 111 |
+---------------+-----+-----+-----+-----+-----+-----+-----+-----+
| new state     |  0  |  1  |  1  |  1  |  1  |  0  |  0  |  0  |

An important convention to be made here is that bits are numbered and showed
from left to right. In binary representation, triplets 000 through 111 
correspond to numbers 0 through 7.

Installation
------------

To compile bitfield from source code:

    $ autoconf
    $ ./configure --prefix=DIRECTORY
    $ make

The only compile-time and run-time dependency is Vitalie Ciubotaru's 'bitfield' 
library (version 0.4.*). If bitfield is not installed, or if the sources were 
configured with the '--with-local-bitfield' flag, make will assume that the 
source code of bitfield is available in the source tree (in 'bitfield' 
directory) and will attempt to build it first.

`make` builds a shared and a static library. For a shared or a static library only, do `make shared` or `make static`.
`make check` runs some tests agains the newly-built library.
`make examples` runs some examples with graphical output.

Functions
---------
For function syntax, see "rule30.h". For details on every function, see
its manual page.

rule30_parents_del() deletes a parents structure.

rule30_parents_new() creates an empty parents structure.

rule30_rev_bit() returns the four parent triplets of a bit (0 or 1)

rule30_rev_ring() treats the input array of bits as a ring and attempts
to reverse-transform it by Wolfram's Rule 30.

rule30_rev_string() reverse-transforms an array of bits by Wolfram's
Rule 30, and returns all possible parent combinations.

rule30_ring_ip() treats the input array of bits as a ring and transforms
it by Wolfram's Rule 30 "in-place".

rule30_ring() treats the input array of bits as a ring and transforms it
by Wolfram's Rule 30.

rule30_ringify() tests if an array of bits can be treated as a ring of
given length and returns the resulting ring.

rule30_string_ip() transforms an array of bits by Wolfram's Rule 30
"in-place".

rule30_string() transforms an array of bits by Wolfram's Rule 30.

Please, see "examples" directory for working examples.

Licensing
---------

rule30 is free software, and is released under the terms of the GNU General 
Public License version 3 or any later version. Please see the file called 
LICENSE.


CHANGELOG
Changelog

0.6.2
* Fixed memory leaks in rule30_rev_ring(), rule30_ringify, rule30_string() and
  rule30_string_ip().

0.6.1
* Fixed a missing-header bug when building with local bitfield source tree.
* Fixed memory leak in rule30_ring() and rule30_ring_ip().
* Added some magic to makefiles ('make distclean' to also run 'make clean').

0.6.0
* Compatibility with bitfield-0.6.
* Minor fixes.

0.4.0
* Added rule30_rev_ring().
* Changed rule30_ringify() syntaxt (the err msg argument is optional now).
* Switched to new bfcmp() function from bitfield-0.4.*.

0.0.1
* Initial release.
* Available functions: rule30_string(), rule30_string_ip(), rule30_ring(), 
  rule30_ring_ip(), rule30_rev_bit(), rule30_rev_string(), rule30_ringify(), 
  rule30_parents_new(), rule30_parents_del().


tests/test1.c
/**
 * File name: tests/test1.c
 * Project name: rule30, an implementation of  Wolfram's Rule 30 written in C
 * URL: https://github.com/ciubotaru/bitfield
 * Author: Vitalie Ciubotaru <vitalie at ciubotaru dot tk>
 * License: General Public License, version 3 or later
 * Date: February 1, 2016
**/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "rule30.h"

/* Testing rule30_string() */

int main()
{
	int i;			//counter
	int len = 80;
	char *msg = "Testing rule30_string()";
	char *failed = "[FAIL]";
	char *passed = "[PASS]";
	int dots = len - strlen(msg) - 6;	/* 6 is the length of pass/fail string */
	printf("%s", msg);
	for (i = 0; i < dots; i++)
		printf(".");
	char * input_raw =
	    "00000000000000000000000000000000000000010000000000000000000000000000000000000000";
	char *check_raw =
	    "10101010111010011001010000110000100011001000101011000000001100011011101110000101";
	struct bitfield *input = bfnew(strlen(input_raw));
	struct bitfield *check = bfnew(bfsize(input));
	str2bf_ip(check_raw, check);
	struct bitfield *empty = bfnew(1);
	str2bf_ip("0", empty);
	str2bf_ip(input_raw, input);
	for (i = 0; i < 50; i++) {
		struct bitfield *tmp = bfcat(empty, input, empty);	// 1 + 80 + 1 = 82
		struct bitfield *tmp2 = rule30_string(tmp);
		bfcpy(tmp2, input);
		bfdel(tmp);
		bfdel(tmp2);
	}

	int result = bfcmp(input, check, NULL);
	bfdel(empty);
	bfdel(input);
	bfdel(check);
	if (result != 0) {
		printf("%s\n", failed);
		return 1;
	}
	printf("%s\n", passed);
	return 0;
}


tests/test2.c
/**
 * File name: tests/test2.c
 * Project name: rule30, an implementation of  Wolfram's Rule 30 written in C
 * URL: https://github.com/ciubotaru/bitfield
 * Author: Vitalie Ciubotaru <vitalie at ciubotaru dot tk>
 * License: General Public License, version 3 or later
 * Date: February 1, 2016
**/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "rule30.h"

/* Testing rule30_rev_bit() */

int main()
{
	srand((unsigned)time(NULL));
	int i;			//counter
	unsigned int input;
	int len = 80;
	char *msg = "Testing rule30_rev_bit()";
	char *failed = "[FAIL]";
	char *passed = "[PASS]";
	int dots = len - strlen(msg) - 6;	/* 6 is the length of pass/fail string */
	printf("%s", msg);
	for (i = 0; i < dots; i++)
		printf(".");

	for (i = 0; i < 50; i++) {
		input = rand() % 2;
		struct parents *triplet = rule30_rev_bit(input);
		int choice = rand() % 4;
		struct bitfield *output =
		    rule30_string(triplet->parent[choice]);
		rule30_parents_del(triplet);
		if (bfgetbit(output, 0) != input) {
			printf("%s\n", failed);
			bfdel(output);
			return 1;
		}
		bfdel(output);
	}
	printf("%s\n", passed);
	return 0;
}


tests/test3.c
/**
 * File name: tests/test3.c
 * Project name: rule30, an implementation of  Wolfram's Rule 30 written in C
 * URL: https://github.com/ciubotaru/bitfield
 * Author: Vitalie Ciubotaru <vitalie at ciubotaru dot tk>
 * License: General Public License, version 3 or later
 * Date: February 1, 2016
**/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "rule30.h"

/* Testing rule30_ring() */

int main()
{
	int i;			//counter
	int len = 80;
	char *msg = "Testing rule30_ring()";
	char *failed = "[FAIL]";
	char *passed = "[PASS]";
	int dots = len - strlen(msg) - 6;	/* 6 is the length of pass/fail string */
	printf("%s", msg);
	for (i = 0; i < dots; i++)
		printf(".");

	char *input_raw =
	    "00000000000000000000000000000000000000010000000000000000000000000000000000000000";
	char *check_raw =
	    "11011001011010011001010000110000100011001000101011000000001100011011101110010010";
	struct bitfield *input = bfnew_quick(len);
	struct bitfield *check = bfnew_quick(len);
	str2bf_ip(input_raw, input);
	str2bf_ip(check_raw, check);
	for (i = 0; i < 50; i++) {
		struct bitfield *tmp = rule30_ring(input);
		bfcpy(tmp, input);
		bfdel(tmp);
	}
	int result = bfcmp(input, check, NULL);
	bfdel(input);
	bfdel(check);
	if (result != 0) {
		printf("%s\n", failed);
		return 1;
	}
	printf("%s\n", passed);
	return 0;
}


rule30.h
/**
 * File name: rule30.h
 * Project name: rule30, an implementation of  Wolfram's Rule 30 written in C
 * URL: https://github.com/ciubotaru/bitfield
 * Author: Vitalie Ciubotaru <vitalie at ciubotaru dot tk>
 * License: General Public License, version 3 or later
 * Copyright 2015, 2016
**/

#ifndef RULE30_H
#define RULE30_H
#include <bitfield.h>

struct parents {
	struct bitfield *parent[4];	// a pointer to an array of 4 struct bitfield
};

void rule30_parents_del(struct parents *instance);

struct parents *rule30_parents_new(const int size);

struct parents *rule30_rev_bit(const int input);	/* returns the four parent triplets of a bit (0 or 1) */

struct parents *rule30_rev_ring(const struct bitfield *input, int *count);	/* treat the input array of bits as a ring and attempt to reverse-transform it by Wolfram's Rule 30 */

struct parents *rule30_rev_string(const struct bitfield *input);	/* reverse-transforms an array of bits by Wolfram's Rule 30, and returns all possible parent combinations */

void rule30_ring_ip(struct bitfield *instance);	/* treats the input array of bits as a ring and transforms it by Wolfram's Rule 30 "in-place" */

struct bitfield *rule30_ring(const struct bitfield *input);	/* treats the input array of bits as a ring and transforms it by Wolfram's Rule 30 */

int rule30_ringify(const struct bitfield *input, struct bitfield *output, char **errmsg);	/* tests if an array of bits can be treated as a ring of given length and returns the resulting ring */

void rule30_string_ip(struct bitfield *instance);	/* transforms an array of bits by Wolfram's Rule 30 "in-place" */

struct bitfield *rule30_string(const struct bitfield *input);	/* transforms an array of bits by Wolfram's Rule 30 */

#endif


rule30.c
/**
 * File name: rule30.c
 * Project name: rule30, an implementation of  Wolfram's Rule 30 written in C
 * URL: https://github.com/ciubotaru/bitfield
 * Author: Vitalie Ciubotaru <vitalie at ciubotaru dot tk>
 * License: General Public License, version 3 or later
 * Copyright 2015, 2016
**/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "rule30.h"

void rule30_parents_del(struct parents *instance)
{
	int i;
	for (i = 0; i < 4; i++)
		bfdel(instance->parent[i]);
	free(instance);
	return;
}

struct parents *rule30_parents_new(const int size)
{
	/* returns a pointer to struct parent */
	int i;
	struct parents *instance = malloc(4 * sizeof(struct parents));
	for (i = 0; i < 4; i++)
		(instance->parent)[i] = bfnew(size);
	return instance;
}

struct parents *rule30_rev_bit(const int input)
{
	struct parents *output = rule30_parents_new(3);
	if (input & 1) {
		str2bf_ip("100", output->parent[0]);
		str2bf_ip("010", output->parent[1]);
		str2bf_ip("011", output->parent[2]);
		str2bf_ip("001", output->parent[3]);
	} else {
		str2bf_ip("000", output->parent[0]);
		str2bf_ip("101", output->parent[1]);
		str2bf_ip("110", output->parent[2]);
		str2bf_ip("111", output->parent[3]);
	}
	return output;
}

struct parents *rule30_rev_ring(const struct bitfield *input, int *count)
{
	int i;
	int result;
	int size = bfsize(input);
	struct parents *potential_parents = rule30_rev_string(input);
	struct parents *parents = rule30_parents_new(size);
	struct bitfield *potential_parent = bfnew(size);
	(*count) = 0;
	for (i = 0; i < 4; i++) {
		result =
		    rule30_ringify(potential_parents->parent[i],
				   potential_parent, NULL);
		if (result == 0) {
			struct bitfield *tmp = bfshift(potential_parent, -1);
			bfcpy(tmp, parents->parent[(int)*count]);
			bfdel(tmp);
			(*count)++;
		}
	}
	rule30_parents_del(potential_parents);
	bfdel(potential_parent);
	return parents;
}

struct parents *rule30_rev_string(const struct bitfield *input)
{
	/**
	 * A string is always Rule30-wise reversible.
	 * There are always 4 strings of length n+2
	 * that generate a particular string length n
	 */

	int size = bfsize(input);
	int i, j, k, l;		//counters
	int t_l;		// counter for matches
	int bit;
	struct parents *output = rule30_parents_new(size + 2);
	struct parents *temp_0;
	struct parents *temp_1;

	//get all proto (3 bit) for the first bit of input string
	temp_0 = rule30_rev_bit(bfgetbit(input, 0));

	if (bfsize(input) == 1) {
		for (i = 0; i < 4; i++)
			bfcpy(temp_0->parent[i], output->parent[i]);
		rule30_parents_del(temp_0);
		return output;
	}
	//initialize temp storage arrays
	struct parents *temp_l = rule30_parents_new(size + 2);
	struct parents *temp_r = rule30_parents_new(size + 2);

	for (i = 0; i < 4; i++) {
		bfcpy(temp_0->parent[i], temp_l->parent[i]);
	}
	rule30_parents_del(temp_0);

	for (i = 1; i < bfsize(input); i++) {
		bit = bfgetbit(input, i);
		temp_1 = rule30_rev_bit(bit);
		t_l = 0;
		for (k = t_l; k < 4; k++) {	//4 substrings in temp_l
			for (j = 0; j < 4; j++) {	//4 substrings in temp_r
				struct bitfield *sub1 = bfsub(temp_l->parent[j], i, i + 2);
				struct bitfield *sub2 = bfsub(temp_1->parent[k], 0, 2);
				int result = bfcmp(sub1, sub2, NULL);
				bfdel(sub1);
				bfdel(sub2);
				if (result == 0) {
					// add to potential ancestors, move values of next_bit to prev_bit and move to next bit
					bfcpy(temp_l->parent[j],
					      temp_r->parent[t_l]);

					if (bfgetbit(temp_1->parent[k], 2))
						bfsetbit(temp_r->parent[t_l],
							 i + 2);
					else
						bfclearbit(temp_r->parent[t_l],
							   i + 2);
					t_l++;	// there can be only 4 matches, so stop comparing after the fourth match
				}
			}
		}
		rule30_parents_del(temp_1);
		for (l = 0; l < 4; l++) {
			bfcpy(temp_r->parent[l], temp_l->parent[l]);
		}
	}
	for (i = 0; i < 4; i++) {
		bfcpy(temp_r->parent[i], output->parent[i]);
	}

	rule30_parents_del(temp_l);
	rule30_parents_del(temp_r);
	return output;
}

void rule30_ring_ip(struct bitfield *instance)
{
	struct bitfield *left = bfshift(instance, 1);
	struct bitfield *right = bfshift(instance, -1);
	struct bitfield *or = bfor(instance, right);
	struct bitfield *xor = bfxor(left, or);
	bfcpy(xor, instance);
	bfdel(left);
	bfdel(right);
	bfdel(or);
	bfdel(xor);
}

struct bitfield *rule30_ring(const struct bitfield *input)
{
	int input_size = bfsize(input);
	struct bitfield *left = bfshift(input, 1);
	struct bitfield *right = bfshift(input, -1);
	struct bitfield *output = bfnew_quick(input_size);
	struct bitfield *or = bfor(input, right);
	struct bitfield *xor = bfxor(left, or);
	bfcpy(xor, output);
	bfdel(left);
	bfdel(right);
	bfdel(or);
	bfdel(xor);
	return output;
}

int rule30_ringify(const struct bitfield *input, struct bitfield *output,
		   char **errmsg)
{
	/* This function extracts 2 sub bitfields that overlap and compares them */

	char *msg;
	int input_size = bfsize(input);
	int output_size = bfsize(output);
	int diff_bits = input_size - output_size;	// expected to be non-negative, but...

	/* if input and output are of the same length, just copy and return */
	if (diff_bits == 0) {
		bfcpy(input, output);
		return 0;
	}

	/* the resulting ring can not be longer than the input string */
	else if (diff_bits < 0) {
		msg = "Input too short";
		goto error;
	}

	/* multiple overlaps are not implemented */
	else if (output_size < diff_bits) {
		msg = "Input too long";
		goto error;
	}

	int end_front = input_size - output_size;	// the last bit of the first subfield, plus one
	int start_back = output_size;	// the first bit of the second subfield
	struct bitfield *front = bfsub(input, 0, end_front);
	struct bitfield *end = bfsub(input, start_back, input_size);

	/* check if the overlapping parts are identical */
	int result = bfcmp(front, end, NULL);
	bfdel(front);
	bfdel(end);
	if (result != 0) {
		msg = "Can't ringify";
		goto error;
	}
	struct bitfield *sub = bfsub(input, 0, output_size);
	bfcpy(sub, output);
	bfdel(sub);
	return 0;
 error:
	if (errmsg) {
		*errmsg = malloc(strlen(msg) + 1);
		if (*errmsg)
			memcpy(*errmsg, msg, strlen(msg) + 1);
	}
	return 1;
}

void rule30_string_ip(struct bitfield *instance)
{
	int size = bfsize(instance);
	struct bitfield *left = bfsub(instance, 0, size - 2);
	struct bitfield *center = bfsub(instance, 1, size - 1);
	struct bitfield *right = bfsub(instance, 2, size);
	bfresize(instance, size - 2);
	/* compute the child generation by Rule 30:
	 * Child(i) = Parent(i-1) XOR ( Parent(i) OR Parent(i+1) )
	 */
	struct bitfield *or = bfor(center, right);
	struct bitfield *xor = bfxor(left, or);
	bfcpy(xor, instance);
	bfdel(left);
	bfdel(center);
	bfdel(right);
	bfdel(or);
	bfdel(xor);
}

struct bitfield *rule30_string(const struct bitfield *input)
{
	int input_size = bfsize(input);
	struct bitfield *left = bfsub(input, 0, input_size - 2);
	struct bitfield *center = bfsub(input, 1, input_size - 1);
	struct bitfield *right = bfsub(input, 2, input_size);
	struct bitfield *output = bfnew_quick(input_size - 2);
	/* compute the child generation by Rule 30:
	 * Child(i) = Parent(i-1) XOR ( Parent(i) OR Parent(i+1) )
	 */
	struct bitfield *tmp1 = bfor(center, right);
	struct bitfield *tmp2 = bfxor(left, tmp1);
	bfcpy(tmp2, output);
	bfdel(left);
	bfdel(center);
	bfdel(right);
	bfdel(tmp1);
	bfdel(tmp2);
	return output;
}
